corehq/apps/app_manager/models.py

```diff
@@ -5947,7 +5951,101 @@ def save(self, *args, **kwargs):
         super(GlobalAppConfig, self).save(*args, **kwargs)
 
 
+class LatestEnabledAppReleases(models.Model):
+    domain = models.CharField(max_length=255, null=False)
+    app_id = models.CharField(max_length=255, null=False)
+    location = models.ForeignKey(SQLLocation, on_delete=models.CASCADE, to_field='location_id')
+    build_id = models.CharField(max_length=255, null=False)
+    version = models.IntegerField(null=False)
+    active = models.BooleanField(default=True)
+    activated_on = models.DateTimeField(null=True)
+    deactivated_on = models.DateTimeField(null=True)
+
+    @staticmethod
+    def expire_cache(domain, location_id, app_id):
+        """
+        expire cache for all location and its descendants for the app
+        why? : Latest enabled release for a location is dependent on restrictions added for
+        itself and its ancestors. Hence we expire the cache for location and its descendants for which the
+        latest enabled release would depend on this location
+        :param domain: domain name
+        :param location_id: location for which the restriction is being updated
+        :param app_id: app id
+        """
+        location = SQLLocation.active_objects.get(location_id=location_id)
+        location_and_descendats = location.get_descendants(include_self=True)
+        for loc in location_and_descendats:
+            get_latest_enabled_app_release.clear(domain, loc.location_id, app_id)
+
+    def clean(self):
+        enabled_release = get_latest_enabled_app_release(self.domain, self.location.location_id, self.app_id)
+        if enabled_release:
+            if enabled_release.version > self.version:
+                raise ValidationError({'version': _("Higher version {} already enabled for this application and "
+                                                    "location").format(enabled_release.version)})
+
+    @classmethod
+    def update_status(cls, domain, app_id, build_id, location_id, version, status):
+        """
+        create a new object or just set the status of an existing one with provided
+        domain, app_id, build_id, location_id and version to the status passed
+        :param domain: domain name
+        :param app_id: app id
+        :param build_id: id of the build corresponding to the version
+        :param location_id: location_id of location
+        :param version: version number
+        :param status: expected status
+        """
+        try:
+            enabled_app_release = LatestEnabledAppReleases.objects.get(
```

Comment:
you could use `get_or_create` with `defaults` here

Reply:
I was using get_or_create here but then updated to be able to use model validation itself in this [commit](https://github.com/dimagi/commcare-hq/pull/23680/commits/e61920e2328af091667f47d78408af01af3546ea)

this also helps me avoid a create and then update, though one can say that that scenario can be avoided by adding `auto_now_add` to `activated_on` because the view does not allow setting active=False and hence the default value is picked always(as of now). 
But then the custom validation i have added would not run as a part of `get_or_create` and i would need to handle that case now after creation. Hence chose to do it this way.